main/app.c

```diff
@@ -2306,16 +2541,219 @@ int ast_app_group_update(struct ast_channel *old, struct ast_channel *new)
+			break; /* We only have one list item per group@category */
+		}
+	}
+	AST_RWLIST_TRAVERSE_SAFE_END;
+
+	if (!gmi) {
+		if (!(gmi = ast_calloc(1, sizeof(struct ast_group_meta)))) {
+		    AST_RWLIST_UNLOCK(&groups_meta);
+		    return -1;
+		}
+
+		ast_copy_string(gmi->group, group, MAX_GROUP_LEN);
+
+		if (!ast_strlen_zero(category)) {
+			ast_copy_string(gmi->category, category, MAX_CATEGORY_LEN);
+		}
+
+		AST_RWLIST_INSERT_TAIL(&groups_meta, gmi, group_meta_list);
+
+		manager_event(EVENT_FLAG_DIALPLAN, "GroupCreate",
+			"Category: %s\r\n"
+			"Group: %s\r\n",
+			category, group);
+	}
+
+	gmi->num_channels++;
+
+	AST_RWLIST_UNLOCK(&groups_meta);
+
+	manager_event(EVENT_FLAG_DIALPLAN, "GroupChannelAdd",
+		"Channel: %s\r\n"
+		"Category: %s\r\n"
+		"Group: %s\r\n"
+		"Uniqueid: %s\r\n",
+		ast_channel_name(chan),
+		category, group,
+		ast_channel_uniqueid(chan));
+
+	return 1;
+}
+
+/* Remove all channels from the given group */
+int ast_app_group_remove_all_channels(const char *group, const char *category)
+{
+	struct ast_group_info *gi = NULL;
+
+	int channels_found = 0;
+
+	if (!category) {
+		category = "";
+	}
+
+	/* Traverse group@category channel assignments */
+	AST_RWLIST_WRLOCK(&groups);
+	AST_RWLIST_TRAVERSE_SAFE_BEGIN(&groups, gi, group_list) {
+		if (strcasecmp(gi->group, group) || strcasecmp(gi->category, category)) {
+			/* Need to match group@category exactly */
+			continue;
+		}
+
+		ast_app_group_remove_channel(gi->chan, gi->group, gi->category);
+		AST_RWLIST_REMOVE_CURRENT(group_list);
+		ast_free(gi);
+
+		channels_found++;
+	}
+	AST_RWLIST_TRAVERSE_SAFE_END;
+	AST_RWLIST_UNLOCK(&groups);
+
+	return !(channels_found > 0);
+}
+
+int ast_app_group_rename(const char *old_group, const char *old_category, const char *new_group, const char *new_category)
+{
+	struct ast_group_info *gi = NULL;
+	struct ast_group_meta *gmi = NULL; /*!< Group metadatas	  */
+
+	int channels_found = 0;
+
+	if (!old_category) {
+		old_category = "";
+	}
+
+	if (!new_category) {
+		new_category = "";
+	}
+
+	/* Traverse group@category channel assignments */
+	AST_RWLIST_WRLOCK(&groups);
+	AST_RWLIST_TRAVERSE_SAFE_BEGIN(&groups, gi, group_list) {
+		if (strcasecmp(gi->group, old_group) || strcasecmp(gi->category, old_category)) {
+			/* Need to match group@category exactly */
+			continue;
+		}
+
+		ast_copy_string(gi->group,    new_group,    MAX_GROUP_LEN);
+		ast_copy_string(gi->category, new_category, MAX_CATEGORY_LEN);
+		channels_found++;
+	}
+	AST_RWLIST_TRAVERSE_SAFE_END;
+	AST_RWLIST_UNLOCK(&groups);
+
+
+	/* Group Variables */
+	AST_RWLIST_WRLOCK(&groups_meta);
+	AST_RWLIST_TRAVERSE_SAFE_BEGIN(&groups_meta, gmi, group_meta_list) {
+		if (strcasecmp(gmi->group, old_group) || strcasecmp(gmi->category, old_category)) {
+			/* Need to match group@category exactly */
+			continue;
+		}
+
+		strncpy(gmi->group,    new_group,    MAX_GROUP_LEN    - 1);
+		strncpy(gmi->category, new_category, MAX_CATEGORY_LEN - 1);
+		channels_found++;
```

Comment:
This shouldn't increment the channel count since it's for a group, not a channel.

Reply:
Fixed